208 iquery scraper: perform a fixed sweep while the court hasn't caught up

albertisfu opened the pull request:
This PR implements the improvements suggested in https://github.com/freelawproject/infrastructure/issues/208#issuecomment-2790685905

The idea is to speed up the iquery scraper for all courts that haven't caught up yet.

To do this, it uses the latest `pacer_case_id` we have in the database for each court to determine whether the scraper is still behind.

If, for a given court, there is a case with a `pacer_case_id` greater than the current `highest_known_pacer_case_id` stored in Redis, we can conclude that the court hasn't caught up. Instead of using the regular exploration mode, the scraper switches to fixed sweep mode.

In fixed sweep mode, the scraper processes `IQUERY_FIXED_SWEEP` (default is 100) case IDs per cycle and updates the `highest_known_pacer_case_id` incrementally. This ensures consistent progress until `highest_known_pacer_case_id` + `IQUERY_FIXED_SWEEP` reaches or exceeds `latest_known_case_id_db`.

Including `IQUERY_FIXED_SWEEP` in the comparison is important to ensure that no case IDs are missed during the transition back to exploration mode.

During fixed mode, probing is not strictly necessary, but we perform one probe just to ensure that the court is not down or hasn’t blocked us.

My initial approach was to retrieve the latest pacer_case_ids directly from the database each time the daemon started. That query leveraged the `pacer_case_id` index to get the largest value per court.

However, when testing this process in the development database, I noticed something strange:
Almost all courts had very large `pacer_case_ids` like 9999, 99999, etc.


```
Latest PACER case IDs by court:
uscfc: 9999
cit: 9999
almb: 99999
alnb: 99999
alsb: 99999
akb: 9999
arb: null
areb: 99999
arwb: 99999
cacb: undefined
caeb: 99999
canb: 99999
casb: 99999
cob: 99999
ctb: 99999
deb: 99999
dcb: 9999
flmb: 999999
flnb: 99999
flsb: 99999
gamb: 99999
ganb: 999999
gasb: 99999
hib: 9999
idb: 99999
ilcb: 99999
ilnb: null
ilsb: 99999
innb: 99999
insb: 99999
ianb: 99999
iasb: 99999
ksb: 99999
kyeb: 99999
kywb: 99999
laeb: 99999
lamb: 9999
lawb: 99999
meb: 99999
mdb: 99999
mab: 99999
mieb: 999999
miwb: 99999
mnb: 99999
msnb: 99999
mssb: 99999
moeb: 99999
mowb: 99999
mtb: 9999
nebraskab: 99999
nvb: 99999
nhb: 99999
njb: 999999
nmb: 99999
nyeb: 99999
nynb: 99999
nysb: 99999
nywb: 99999
nceb: 99999
ncmb: 99999
ncwb: 99999
ndb: 9999
ohnb: 999999
ohsb: 99999
okeb: 9999
oknb: 99999
okwb: 99999
orb: 99999
paeb: 99999
pamb: 99999
pawb: 99999
rib: 99999
scb: 99999
sdb: 9999
tneb: 99999
tnmb: 99999
tnwb: 99999
txeb: 99999
txnb: 99999
txsb: 99999
txwb: 99999
utb: 9999
vtb: 9999
vaeb: null
vawb: 99999
waeb: 99999
wawb: 99999
wvnb: 9999
wvsb: 99999
wieb: 99999
wiwb: null
wyb: 9999
gub: null
nmib: 98
prb: 99999
vib: 999
dcd: 99999
almd: 9999
alnd: 99999
alsd: 9999
akd: 9999
azd: 999999
ared: 99999
arwd: 9999
cacd: 99999
caed: 99999
cand: 99999
casd: 99999
cod: 99999
ctd: 99999
ded: 9999
flmd: 99999
flnd: 99998
flsd: 99999
gamd: 99999
gand: 99999
gasd: 9999
hid: cwdatty
idd: 9999
ilcd: 9999
ilnd: 99999
ilsd: 99999
innd: 99999
insd: 99999
iand: 9999
iasd: 9999
ksd: 99999
kyed: 99999
kywd: 99999
laed: 99999
lamd: 9999
lawd: 99999
med: 9999
mdd: 99999
mad: 99999
mied: 99999
miwd: 99999
mnd: 99999
msnd: 9999
mssd: 99999
moed: 99999
mowd: 99999
mtd: 9999
ned: 99999
nvd: termall
nhd: 9999
njd: 99999
nmd: 99999
nyed: 99999
nynd: 99999
nysd: 99999
nywd: 99999
nced: 99999
ncmd: 99999
ncwd: 99998
ndd: 9999
ohnd: 99999
ohsd: 99999
oked: 9999
oknd: 9999
okwd: 99999
ord: 99999
paed: 99999
pamd: 99999
pawd: 99999
rid: 9999
scd: 99999
sdd: 9999
tned: 99999
tnmd: 99999
tnwd: 99999
txed: 99994
txnd: 99999
txsd: 999999
txwd: 999999
utd: 99999
vtd: 9999
vaed: 99999
vawd: 99999
waed: 99994
wawd: 99999
wvnd: 9999
wvsd: 99999
wied: 99999
wiwd: 9999
wyd: 9999
gud: 9999
nmid: 999
prd: 99998
vid: 9998
jpml: 999999

Time taken: 20.3592 seconds
```
 
I reviewed some of these cases, such as:
https://www.courtlistener.com/docket/61315948/united-states-v-prado-lerma/

The `pacer_case_id` 999999 does exist in PACER, but it's an old case from 2012.
So this might be a bug on PACER’s end.

Therefore, it's not reliable to simply use the largest `pacer_case_id` for each court.

Instead, it's better to rely on the `date_filed`, so we can retrieve the most recent case we have in the database as of the day the case IDs are set.

However, since `date_filed` is not an indexed field, these queries are quite slow. I estimate it would take a couple of hours to retrieve the IDs from all courts. For that reason, it’s not a good idea to run them every time the daemon is restarted. Instead, I moved that logic to a separate command so the IDs can be stored in Redis.

Once this is merged, the command can be run as:

`manage.py ready_mix_cases_project --task set-latest-case-ids --court-type all --database default
`

After these IDs are stored, the daemon will begin using them whenever the current `highest_known_pacer_case_id` is smaller than the latest known `pacer_case_id`.


Other tweak we can do after the command above is run.

Reduce `IQUERY_PROBE_WAIT` environment variable from `14_400` to `7_200`, since we'll now process 100 cases per cycle. This allows us to decrease the wait time between cycles and monitor how Elasticsearch behaves. If performance permits, we can reduce it further.

mlissner commented:
Looks and sounds about right to me. Bummer about date_filed not being indexed. Onward for a full review.

ERosendo commented on cl/corpus_importer/tasks.py:
```diff
@@ -1449,11 +1454,39 @@ def probe_iquery_pages(
+    probe_iteration_limit = (
+        1 if do_fixed_sweep else settings.IQUERY_PROBE_MAX_OFFSET
+    )
+    while probe_offset + jitter < probe_iteration_limit:
```
we're setting `probe_iteration_limit` to 1 when doing a `fixed_sweep` just to make sure we haven't been blocked or the court is not down

The `while` loop condition checks if `offset + jitter` is less than `iteration_limit`. Given that `jitter` is a random value, it could be bigger than 1. Is this the intended behavior? If jitter is greater than 0, we won't check court availability.

albertisfu replied:
Thanks for catching this. You're right this is not the intended behaviour. I've updated the code to set the jitter to 0 if `do_fixed_sweep` is True.

ERosendo reviewed (approved):
The code looks good. Just one comment on the validation in the `probe_iquery_pages` task. After that's resolved, it's good to merge.

albertisfu commented:
Thanks @ERosendo, I've addressed your comment. I also restored the original `probe_iquery_pages` task since the signature of the updated task has changed. This could cause tasks already in the queue to fail. It's better to keep the original task for now and remove it after this is deployed and a few hours have passed, ensuring the old tasks in the queue have been processed.
